Add logger specific to API (#2086)

hoenirvili opened the pull request:
## Motivation
All the **grpc** common functionality are placed in a single package and their all using the global logger. This is not good because what if the client want's to use a custom logger or other miscellaneous options?
 
## Closes #2086 

## Changes
I've changed all files under api/grpcserver/* to make all the concrete implementation be flexible. By flexible I mean, using the option pattern and make all implementations support adding required miscellaneous things like support custom loggin, support grpc protocol options etc. Because I didn't wanted types like **optionNode**, **optionMesh**, etc. I've moved every concrete implementation into it's own package under grpcserver/* to make the naming once again sane and be more maintainable.


## Test Plan
<!-- Please specify how these changes were tested 
(e.g. unit tests, manual testing, etc.) -->

I've just re-run the unit tests and made sure that everything passes alongside with whole compilation of the main binary (on linux 64bit)

## TODO
<!-- This section should be removed when all items are complete -->
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

lrettig committed:
Some minor cleanup

lrettig committed:
Upgrade to latest zap module

lrettig committed:
Fix log levels issue

Simplify the wrapped Log type
Set level of derived logger correctly, and warn if the level is too low

lrettig committed:
Clean up some misc log statements

y0sher commented:
Hi @hoenirvili , the whole thing looks great. I would really like @lrettig to have a look since he is working on the API atm. 
also one thing I can tell is that this PR is missing some functionality we have for loggers - controlling the log level dynamically. 
if you look on the top of the `node.go` flie you'd find constants we use to name each log, this is created by maintaining a map with the state and loading it from config or changed from the grpc api. please have a look and try to add it, contact me if you need any help.

noamnelke commented on cmd/node/node.go:
```diff
@@ -48,9 +52,11 @@ import (
 	timeCfg "github.com/spacemeshos/go-spacemesh/timesync/config"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+
+	_ "net/http/pprof"
 )
 
-import _ "net/http/pprof" // import for memory and network profiling
+// import for memory and network profiling
```
the comment should be on the same line as the import - it explains it

noamnelke replied:
note that this also causes linting to fail

hoenirvili replied:
Thanks, I did move the comment back where it was. I think go-fmt moved the single **import** into the big import and didn't moved the comment.

noamnelke reviewed (commented):
some of the file renames were correctly understood by git and others weren't - can you please fix this so it's easier to see the changes in the moved files, and to maintain the correct history of the files?

noamnelke commented:
Thinking about this a bit more, I'm not sure the relatively big refactor and partition into packages should be in the same PR as the logger change. @lrettig ?

noamnelke commented:
Note that `TestSpacemeshApp_GrpcService` panics.

hoenirvili commented:
> 
> 
> Hi @hoenirvili , the whole thing looks great. I would really like @lrettig to have a look since he is working on the API atm.
> also one thing I can tell is that this PR is missing some functionality we have for loggers - controlling the log level dynamically.
> if you look on the top of the `node.go` flie you'd find constants we use to name each log, this is created by maintaining a map with the state and loading it from config or changed from the grpc api. please have a look and try to add it, contact me if you need any help.

Ok, I think I understood what you mean. I did checked the code and saw the map of loggers and I saw that this is configurable and you can set dynamically the log level across multiple resources so I think I managed to add support/port the functionality to all services/servers types.

Can you check one more time and let me know if this is what you want?

y0sher commented on cmd/node/node.go:
```diff
@@ -689,20 +695,30 @@ func (app *SpacemeshApp) startAPIServices(postClient api.PostAPI, net api.Networ
-		startService(grpcserver.NewMeshService(net, app.mesh, app.clock, app.syncer))
+		lvl := app.loggers[MeshLogger] // because we already set the logger in the map
+		startService(servicemesh.New(net, app.mesh, app.clock, app.syncer, servicemesh.WithLogger(
+			lg.SetLevel(lvl).WithName(MeshLogger),
```
didn't you mean to use `addLogger` here as well?

hoenirvili replied:
I've noticed that meshlogger it's already set in the map so I figured I can just get ptr from the map and just pass/construct only the outer required thing. I can change to addLogger but it will overwrite the same mesh logger or we could add a new var key with a different name. You're call.

lrettig replied:
These are definitely two different things - `MeshLogger` should be distinct from `MeshServiceLogger`, which is what you want here. The GRPC services should get distinct loggers with the name `<Svcname>ServiceLogger`.

hoenirvili replied:
I did used/updated all naming logger acordignly.

y0sher commented on cmd/node/node.go:
```diff
@@ -689,20 +695,30 @@ func (app *SpacemeshApp) startAPIServices(postClient api.PostAPI, net api.Networ
 	// Make sure we only start the server once
-	startService := func(svc grpcserver.ServiceAPI) {
+	name := app.nodeID.ShortString()
+	lg := log.NewDefault(name).WithFields(app.nodeID)
```
the method that calls `startAPIServices` (= `Start`) is creating a logger that way already. it will be better to pass it down to this instead of creating a new one. though if you see a lot of clashes in apis and other usages than we can live with this duplication. though I'm pretty sure we only use it once so should be easy to pass down the root logger..

hoenirvili replied:
Ok, make sense. I will reuse the app.Logger I thought by using the same initial constructed one would not work. Thanks for the valuable insight.

lrettig replied:
> Btw, I have a proposal, instead of relaying on that switch stmt why won't we just have a map of these keys like `map[keys]bool` or `map[keys]struct{}` and if an element is found then it will return ok and we can go and use the **logger**. This way we only have to modify in one place the **map** and everything will work and also we also have the benefit at runtime to add dynamic support for _future keys_.
> 
> Do you want me in this **PR** to add this or make a new one after this is merged and continue further improving this ?

I don't have an opinion on replacing the switch with a map, will defer to @y0sher and @noamnelke on that. But if we do make that change, let's do it in a separate branch/PR - small PRs are good! :)

hoenirvili replied:
Ok, sounds good to me !

y0sher commented:
> > Hi @hoenirvili , the whole thing looks great. I would really like @lrettig to have a look since he is working on the API atm.
> > also one thing I can tell is that this PR is missing some functionality we have for loggers - controlling the log level dynamically.
> > if you look on the top of the `node.go` flie you'd find constants we use to name each log, this is created by maintaining a map with the state and loading it from config or changed from the grpc api. please have a look and try to add it, contact me if you need any help.
> 
> Ok, I think I understood what you mean. I did checked the code and saw the map of loggers and I saw that this is configurable and you can set dynamically the log level across multiple resources so I think I managed to add support/port the functionality to all services/servers types.
> 
> Can you check one more time and let me know if this is what you want?

Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.

hoenirvili commented:
> Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.

I addressed the comments, how is it now?

y0sher commented:
> > Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.
> 
> I addressed the comments, how is it now?

The logger attached to `app` is already named and is used for app logs. I'd rather we pass down the root logger created in the function that calls `startAPIServices`.
also sorry I missed it earlier but for the log level config to work you will also have to add config entries for your loggers and add them in `addLogger`.

hoenirvili commented:
> 
> 
> > > Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.
> > 
> > 
> > I addressed the comments, how is it now?
> 
> The logger attached to `app` is already named and is used for app logs. I'd rather we pass down the root logger created in the function that calls `startAPIServices`.
> also sorry I missed it earlier but for the log level config to work you will also have to add config entries for your loggers and add them in `addLogger`.

Thanks once again for the patience, I just modified the code. How is it now?

Btw, I have a proposal, instead of relaying on that switch stmt why won't we just have a map of these keys like ```map[keys]bool``` or ```map[keys]struct{}``` and if an element is found then it will return ok and we can go and use the **logger**. This way we only have to modify in one place the **map** and everything will work and also we also have the benefit at runtime to add dynamic support for *future keys*. 

Do you want me in this **PR** to add this or make a new one after this is merged and continue further improving this ?

lrettig commented on api/pb/api.proto:
```diff
@@ -1,6 +1,8 @@
 syntax = "proto3";
 package pb;
 
+option go_package="api/pb";
```
This file, and all the included protobuf files, will be deprecated very soon - let's not bother updating them, okay?

hoenirvili replied:
Ok, I didn't know that. The reason I added these line to get rid of the annoying warning from generating the proto files.

lrettig replied:
No worries! The files are going away because we have a full golang build of the new API in https://github.com/spacemeshos/api/tree/master/release/go, btw, and we can use it directly as a library rather than compiling the protobuf inside go-spacemesh, which is much cleaner (and better separation of concerns).

lrettig commented on cmd/node/node.go:
```diff
@@ -80,6 +83,10 @@ const (
 	NipstBuilderLogger   = "nipstBuilder"
 	AtxBuilderLogger     = "atxBuilder"
+	GrpcLogger           = "grpc"
+	NodeLogger           = "node"
```
Let's pick a consistent naming convention for the loggers. I like `meshService` for the mesh service, how about `nodeService` for this one?

hoenirvili replied:
Deal.

lrettig commented on cmd/node/node.go:
```diff
@@ -150,8 +157,8 @@ type SpacemeshApp struct {
 	jsonAPIService    *api.JSONHTTPServer
-	newgrpcAPIService *grpcserver.Server
-	newjsonAPIService *grpcserver.JSONHTTPServer
+	newgrpcAPIService *server.Server
```
It's not as immediately obvious from the package name `server` what this package does, `grpcserver` is a bit clearer and more intuitive isn't it?

hoenirvili replied:
Ok, I will change this into `grpcserver.Server`

lrettig reviewed (changes requested):
Thanks for the PR! Added a few quick comments. GitHub's diff looks off for the renamed files, I will take a closer look at those later. Could you rebase against the `api-phase9` branch? We'll be merging that branch soon, and this will prevent some ugly merge conflicts when that happens. #2092 is also in flight and will land soon and might cause some conflicts here, so we might want to merge that in here as well to prevent downstream conflicts.

lrettig commented:
> Thinking about this a bit more, I'm not sure the relatively big refactor and partition into packages should be in the same PR as the logger change. @lrettig ?

Agree here, it would be nice to split up these concerns - the partitioning into packages isn't required for the logging is it?

lrettig commented:
> some of the file renames were correctly understood by git and others weren't - can you please fix this so it's easier to see the changes in the moved files, and to maintain the correct history of the files?

I'm seeing this issue now as well, odd - @hoenirvili is this something you might be able to fix? I'm not sure why GH does this.

y0sher commented:
> > > > Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.
> > > 
> > > 
> > > I addressed the comments, how is it now?
> > 
> > 
> > The logger attached to `app` is already named and is used for app logs. I'd rather we pass down the root logger created in the function that calls `startAPIServices`.
> > also sorry I missed it earlier but for the log level config to work you will also have to add config entries for your loggers and add them in `addLogger`.
> 
> Thanks once again for the patience, I just modified the code. How is it now?
> 
> Btw, I have a proposal, instead of relaying on that switch stmt why won't we just have a map of these keys like `map[keys]bool` or `map[keys]struct{}` and if an element is found then it will return ok and we can go and use the **logger**. This way we only have to modify in one place the **map** and everything will work and also we also have the benefit at runtime to add dynamic support for _future keys_.
> 
> Do you want me in this **PR** to add this or make a new one after this is merged and continue further improving this ?

Good job on fixing it. the only part missing are the entries in the `config.toml` file for future reference. 
It would be great if you could try and improve this feature but I believe it is better to do so in a separate PR.

hoenirvili commented:
> 
> 
> > > > > Exactly what I wanted, added a few minor comments. still waiting for @lrettig to take a look here before merging.
> > > > 
> > > > 
> > > > I addressed the comments, how is it now?
> > > 
> > > 
> > > The logger attached to `app` is already named and is used for app logs. I'd rather we pass down the root logger created in the function that calls `startAPIServices`.
> > > also sorry I missed it earlier but for the log level config to work you will also have to add config entries for your loggers and add them in `addLogger`.
> > 
> > 
> > Thanks once again for the patience, I just modified the code. How is it now?
> > Btw, I have a proposal, instead of relaying on that switch stmt why won't we just have a map of these keys like `map[keys]bool` or `map[keys]struct{}` and if an element is found then it will return ok and we can go and use the **logger**. This way we only have to modify in one place the **map** and everything will work and also we also have the benefit at runtime to add dynamic support for _future keys_.
> > Do you want me in this **PR** to add this or make a new one after this is merged and continue further improving this ?
> 
> Good job on fixing it. the only part missing are the entries in the `config.toml` file for future reference.
> It would be great if you could try and improve this feature but I believe it is better to do so in a separate PR.


I just did that now, I still need to figure out how to sync with `api-phase9` and it should be fine after that I think.

lrettig commented:
> I just did that now, I still need to figure out how to sync with api-phase9 and it should be fine after that I think.

Step one: create a backup of your current branch in case the rebase goes wrong :)

`> git checkout -b mybranch-backup && git checkout -`

Step two: rebase against api-phase9. Hopefully this is relatively painless.

`> git rebase origin/api-phase9`

Step three: force push your changes

`> git push -f`

Step four: [change the base branch](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request) of this PR.

Some more possibly helpful info here:

- https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request
- https://anavarre.net/how-to-rebase-a-github-pull-request/
- https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

hoenirvili commented:
> 
> 
> > I just did that now, I still need to figure out how to sync with api-phase9 and it should be fine after that I think.
> 
> Step one: create a backup of your current branch in case the rebase goes wrong :)
> 
> `> git checkout -b mybranch-backup && git checkout -`
> 
> Step two: rebase against api-phase9. Hopefully this is relatively painless.
> 
> `> git rebase origin/api-phase9`
> 
> Step three: force push your changes
> 
> `> git push -f`
> 
> Step four: [change the base branch](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request) of this PR.
> 
> Some more possibly helpful info here:
> 
>     * https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request
> 
>     * https://anavarre.net/how-to-rebase-a-github-pull-request/
> 
>     * https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

Thanks man you're awesome, tho' I knew what I had to do (as in git commands), just I'm lazy to resolve all the conflicts that appeared after rebase. I will fix this asap tho.'

lrettig commented:
Sounds good. Hopefully there aren't too many conflicts! Sorry to create
extra work for you

On Tue, Aug 4, 2020, 12:55 hœnir <notifications@github.com> wrote:

> I just did that now, I still need to figure out how to sync with
> api-phase9 and it should be fine after that I think.
>
> Step one: create a backup of your current branch in case the rebase goes
> wrong :)
>
> > git checkout -b mybranch-backup && git checkout -
>
> Step two: rebase against api-phase9. Hopefully this is relatively painless.
>
> > git rebase origin/api-phase9
>
> Step three: force push your changes
>
> > git push -f
>
> Step four: change the base branch
> <https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request>
> of this PR.
>
> Some more possibly helpful info here:
>
> * https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request
>
> * https://anavarre.net/how-to-rebase-a-github-pull-request/
>
> * https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request
>
> Thanks man you're awesome, tho' I knew what I had to do (as in git
> commands), just I'm lazy to resolve all the conflicts that appeared after
> rebase. I will fix this asap tho.'
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spacemeshos/go-spacemesh/pull/2100#issuecomment-668710419>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZJWNFORVIBFXFDDO7DZK3R7A4QJANCNFSM4PSSBIYA>
> .
>

lrettig commented:
Thanks for rebasing, I'm sure that was a lot of work. I've begun reviewing your PR but it's a very large refactor and it may take me a few days to review it completely.

hoenirvili committed:
Make all new services have the logger be configurable

Signed-off-by: hoenirvili <hoenirvili@gmail.com>

hoenirvili commented:
What about now? Is the patch fine?

jack17529 committed:
Update .dockerignore (#2123)

Add `/devtools` to `.dockerignore`  
Closes https://github.com/spacemeshos/go-spacemesh/issues/2121


Co-authored-by: Lane Rettig <lanerettig@gmail.com>

y0sher committed:
fixed panic bug on chace eviction. added scenario to test.  (#2125)

## Motivation
When udp message from a new peer would arrive and the udp session state cache was full we try to evict one session. if one of the sessions in the state is older than 24hours we remove it first. this code path was never tested and had called `Close` on the connection after removing it from the map.

Closes #
<!-- `Closes #XXXX, closes #XXXX, ...` links mentioned issues to this PR and automatically closes them when this it's merged -->

## Changes

added test and fixed.

## Test Plan
<!-- Please specify how these changes were tested 
(e.g. unit tests, manual testing, etc.) -->

## TODO
<!-- This section should be removed when all items are complete -->
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

lrettig committed:
Do a better job of sending test transactions (#2127)

## Motivation
Closes #2126 

## Changes
- Fail fast when generating and sending transactions to new test accounts fails
- Pause after sending each tx to give the network a chance to gossip the tx before sending another from the same source account

## Test Plan
N/A, this is an update to tests

lrettig committed:
Merge branch 'develop' into api-phase9

lrettig committed:
Merge branch 'develop' into fix-log-levels3

lrettig committed:
Remove InitSpacemeshLoggingSystem

It's redundant and doesn't do anything

lrettig committed:
Adds some basic low level log tests

Make some data items in log configurable for tests
Test all three levels of loggers

lrettig committed:
WIP: Add a couple more tests

Final test is not passing as setting dynamic log level is not currently
working

lrettig committed:
Remove SetLogLevel

Removes this endpoint from the (old) GRPC API and removes the
corresponding method from App. Remove corresponding interface from API
and don't pass App object into API.

Update app tests accordingly. Add a method to log to allow fully custom
logs to be constructed.

lrettig committed:
Use NewDefault to create new logs

lrettig committed:
Refactor log creation

Remove unused data file options, replace New with NewDefault

lrettig committed:
Remove test to change log level

This functionality was removed

lrettig committed:
Merge branch 'develop' into fix-log-levels3

lrettig committed:
Update modules

lrettig committed:
Make lint happy

lrettig committed:
Make linter even happier

lrettig committed:
Merge branch 'develop' into api-phase9

lrettig committed:
Merge branch 'fix-log-levels3' into api-phase9

lrettig committed:
gitignore

lrettig committed:
Merge branch 'api-phase9' into add-logger-issue-2086

hoenirvili commented:
So what's the right decision, I noticed that `api-phase9` is merged with this branch. Should I point now the PR to `api-phase9`and when `api-phase9` is ready it will be merged to `master` ?

Or do you want me this split this into multiple PR's and start sending one at a time to `develop` ?

lrettig commented:
It looks like this was closed automatically since the base branch was merged into develop. I can't reopen this. We'll need to transfer this over to a new PR against develop.

lrettig commented:
> Or do you want me this split this into multiple PR's and start sending one at a time to develop ?

Yes let's do this! :)
